File: src/algorithms/trees/traversal/diagonal-traversal/sources/tree-diagonal-traversal_test.rs
include!("tree-diagonal-traversal.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<BSTNode>>, right: Option<Box<BSTNode>>) -> Option<Box<BSTNode>> {
        Some(Box::new(BSTNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<BSTNode>> {
        make_node(value, None, None)
    }

    #[test]
    fn test_balanced_7_node_bst() {
        let root = make_node(4,
            make_node(2, leaf(1), leaf(3)),
            make_node(6, leaf(5), leaf(7)));
        assert_eq!(tree_diagonal_traversal(root), vec![vec![4, 6, 7], vec![2, 5, 3], vec![1]]);
    }

    #[test]
    fn test_null_root() {
        assert_eq!(tree_diagonal_traversal(None), Vec::<Vec<i32>>::new());
    }

    #[test]
    fn test_single_node() {
        assert_eq!(tree_diagonal_traversal(leaf(42)), vec![vec![42]]);
    }

    #[test]
    fn test_right_skewed() {
        let root = make_node(1, None, make_node(2, None, leaf(3)));
        assert_eq!(tree_diagonal_traversal(root), vec![vec![1, 2, 3]]);
    }

    #[test]
    fn test_left_skewed() {
        let root = make_node(3, make_node(2, leaf(1), None), None);
        assert_eq!(tree_diagonal_traversal(root), vec![vec![3], vec![2], vec![1]]);
    }
}

File: src/algorithms/trees/traversal/diagonal-traversal/tree-diagonal-traversal_test.py
import importlib
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "sources"))

mod = importlib.import_module("tree-diagonal-traversal")
diagonal_traversal = mod.diagonal_traversal
BSTNode = mod.BSTNode


def make_node(value, left=None, right=None):
    node = BSTNode(value)
    node.left = left
    node.right = right
    return node


def test_balanced_7_node_bst():
    root = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))
    assert diagonal_traversal(root) == [[4, 6, 7], [2, 5, 3], [1]]


def test_null_root():
    assert diagonal_traversal(None) == []


def test_single_node():
    assert diagonal_traversal(make_node(42)) == [[42]]


def test_right_skewed():
    root = make_node(1, None, make_node(2, None, make_node(3)))
    assert diagonal_traversal(root) == [[1, 2, 3]]


def test_left_skewed():
    root = make_node(3, make_node(2, make_node(1)))
    assert diagonal_traversal(root) == [[3], [2], [1]]


if __name__ == "__main__":
    test_balanced_7_node_bst()
    test_null_root()
    test_single_node()
    test_right_skewed()
    test_left_skewed()
    print("All tests passed!")
